core/src/main/java/hudson/util/RobustReflectionConverter.java

```diff
@@ -271,8 +271,71 @@ public Object unmarshal(final HierarchicalStreamReader reader, final Unmarshalli
         return serializationMethodInvoker.callReadResolve(result);
     }
 
+    private static final class FieldExpectation {
+        private final Class definingClass;
+        private final String name;
+
+        public FieldExpectation(Class definingClass, String name) {
+            this.definingClass = definingClass;
+            this.name = name;
+        }
+
+        public Class getDefiningClass() {
+            return definingClass;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            FieldExpectation that = (FieldExpectation) o;
+
+            if (definingClass != null ? !definingClass.equals(that.definingClass) : that.definingClass != null) {
+                return false;
+            }
+            return name.equals(that.name);
+        }
+
+        @Override
+        public int hashCode() {
+            int result = definingClass != null ? definingClass.hashCode() : 0;
+            result = 31 * result + name.hashCode();
+            return result;
+        }
+
+        @Override
+        public String toString() {
+            return "FieldExpectation{" +
+                    "definingClass=" + definingClass +
+                    ", name='" + name + '\'' +
+                    '}';
+        }
+
+
+    }
+
     public Object doUnmarshal(final Object result, final HierarchicalStreamReader reader, final UnmarshallingContext context) {
         final SeenFields seenFields = new SeenFields();
+        final boolean existingObject = context.currentObject() != null;
+        final Map<FieldExpectation, Object> expectedFields = existingObject ? new HashMap<FieldExpectation, Object>() : null;
+        final Object cleanInstance = existingObject ? reflectionProvider.newInstance(result.getClass()) : null;
```

Comment:
`reflectionProvider` always seems to be an instance of `SunLimitedUnsafeReflectionProvider` which uses `sun.misc.Unsafe#allocateInstance` to create a new instance. AFAIK `Unsafe#allocateInstance` does not call any constructor, so fields will not be initialized. So the value of each field in `cleanInstance` will be `null`, `0`, `false`, etc. That's why there are so many NPEs in the tests.